crates/storage/db/src/snapshot.rs

```diff
@@ -0,0 +1,106 @@
+//! reth's snapshot creation from database tables
+
+use crate::{
+    abstraction::cursor::DbCursorRO,
+    table::{Key, Table},
+    transaction::DbTx,
+    DatabaseError, RawKey, RawTable,
+};
+use reth_interfaces::RethResult;
+use reth_nippy_jar::NippyJar;
+use std::ops::{RangeBounds, RangeInclusive};
+
+/// Macro that generates snapshot creation functions that take an arbitratry number of [`Table`] and
+/// creates a [`NippyJar`] file out of their [`Table::Value`]. Each list of [`Table::Value`] from a
+/// table is a column of values.
+///
+/// Has membership filter set and compression dictionary support.
+macro_rules! generate_snapshot_func {
+    ($(($($tbl:ident),+)),+ $(,)? ) => {
+        $(
+            paste::item! {
+                /// Creates a snapshot from specified tables. Each table's `Value` iterator represents a column.
+                ///
+                /// **Ensure the range contains the same number of rows.**
+                ///
+                /// * `tx`: Database transaction.
+                /// * `range`: Data range for columns in tables.
+                /// * `dict_compression_set`: Sets of column data for compression dictionaries. Max size is 2GB. Row count is independent.
+                /// * `row_count`: Total rows to add to `NippyJar`. Must match row count in `range`.
+                /// * `nippy_jar`: Snapshot object responsible for file generation.
+                #[allow(non_snake_case)]
+                pub fn [<create_snapshot$(_ $tbl)+>]<'tx,
+                    $($tbl: Table<Key=K>,)+
+                    HashTbl: Table<Key=K>,
```

Comment:
will it work for transaction-based snapshots? We don't have a table with `TxNumber => TxHash` mapping, because we don't store hashes